### ISSUE TITLE
Empty transactions are rolled back differently in Bolt and HTTP

### ISSUE THREAD

**Issue by fpavageau** [neo4j/neo4j-jdbc#62]
PR fbiville/liquigraph#105 in Liquigraph currently fails when using the HTTP driver but not the Bolt driver because of a different behavior in `Connection.setAutoCommit(false)` leading to a different code path taken in `Connection.rollback()` when the uncommitted transaction is actually empty.
# `rollback` implementation
## HTTP driver

[`HttpConnection`](https://github.com/neo4j-contrib/neo4j-jdbc/blob/3.0/neo4j-jdbc-http/src/main/java/org/neo4j/jdbc/http/HttpConnection.java#L135):

``` java
    @Override public void rollback() throws SQLException {
        checkClosed();
        checkAutoCommit();
        executor.rollback();
    }
```

delegates to [`CypherExecutor`](https://github.com/neo4j-contrib/neo4j-jdbc/blob/3.0/neo4j-jdbc-http/src/main/java/org/neo4j/jdbc/http/driver/CypherExecutor.java#L188):

``` java
    public void rollback() throws SQLException {
        if (this.getOpenTransactionId() > 0) {
            // Prepare the request
            HttpDelete request = new HttpDelete(currentTransactionUrl);
            Neo4jResponse response = this.executeHttpRequest(request);
            if (response.code != 200 & response.hasErrors()) {
                throw new SQLException(response.displayErrors());
            }
            this.currentTransactionUrl = this.transactionUrl;
        } else {
            throw new SQLException("There is no transaction to rollback");
        }
    }
```

A `SQLException` will be thrown when rolling back an empty transaction.
## Bolt driver

[`BoltConnection`](https://github.com/neo4j-contrib/neo4j-jdbc/blob/3.0/neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltConnection.java#L121):

``` java
    @Override public void rollback() throws SQLException {
        this.checkClosed();
        this.checkAutoCommit();
        if (this.transaction == null) {
            throw new SQLException("The transaction is null");
        }
        this.transaction.failure();
    }
```

A `SQLException` will also be thrown when rolling back a null transaction, even if the message is different. However, a transaction can be empty without being null.
# `setAutoCommit` implementation
## HTTP driver

[`HttpConnection`](https://github.com/neo4j-contrib/neo4j-jdbc/blob/3.0/neo4j-jdbc-http/src/main/java/org/neo4j/jdbc/http/HttpConnection.java#L121):

``` java
    @Override public void setAutoCommit(boolean autoCommit) throws SQLException {
        this.executor.setAutoCommit(autoCommit);
    }
```

also delegates to [`CypherExecutor`](https://github.com/neo4j-contrib/neo4j-jdbc/blob/3.0/neo4j-jdbc-http/src/main/java/org/neo4j/jdbc/http/driver/CypherExecutor.java#L214):

``` java
    public void setAutoCommit(Boolean autoCommit) throws SQLException {
        // we only do something if there is a change
        if (this.autoCommit != autoCommit) {

            if (autoCommit) {
                // Check if a transaction is currently opened before
                // If so, we commit it
                if (getOpenTransactionId() > 0) {
                    this.commit();
                }
                this.autoCommit = Boolean.TRUE;
                this.currentTransactionUrl = this.transactionUrl + "/commit";
            } else {
                this.autoCommit = Boolean.FALSE;
                this.currentTransactionUrl = this.transactionUrl;
            }
        }
    }
```

When calling `setAutoCommit(false)` on a connection where `this.autoCommit` is not already `false` (it could be `true` or null, as it's a `Boolean` and not a `boolean`), `this.currentTransactionUrl` is either null or already equal to `this.transactionUrl`, so nothing really happens: the transaction is ready to begin, but not open, there's no transaction id.

_It does not depend on the previous value of `this.autoCommit`._
## Bolt driver

[`BoltConnection`](https://github.com/neo4j-contrib/neo4j-jdbc/blob/3.0/neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltConnection.java#L88):

``` java
    @Override public void setAutoCommit(boolean autoCommit) throws SQLException {
        if (this.autoCommit != autoCommit) {
            if (this.transaction != null && !this.autoCommit) {
                this.commit();
            }

            if (this.autoCommit) {
                //Simply restart the transaction
                this.transaction = this.session.beginTransaction();
            } else {
                this.transaction.close();
            }

            this.autoCommit = autoCommit;
        }
    }
```

_This time, it does depend on the previous value of `this.autoCommit`._

If there's an open transaction and `this.autoCommit` is `false`, the transaction is immediately committed, it's the same behavior as in `CypherExecutor`.

However, when calling `setAutoCommit(false)` and `this.autoCommit` is `true` (it's a `boolean` and not a `Boolean`), as noted in the comment "Simply restart the transaction", the previous transaction is forgotten and a new one is set. `transaction` is **not** set to null, and the `SQLException` will not be triggered if `rollback()` is called directly.

It can also be noted that the fact that `transaction` can be null is handled inconsistently: the `else` clause will always close the transaction, even when it could be null according to the first `if` of the block. Either an extra nullity test is missing, or the existing one is useless.
# Possible fixes
## HTTP driver

It could be modified to immediately begin a transaction in `setAutoCommit(false)`, as is done in the Bolt driver, but at the risk of a timeout if it's kept around too long before being actually used. I think it would become more fragile.
## Bolt driver

It could be modified to manage the transaction lazily, by keeping null and only begin a transaction when starting an interaction with the connection, as is actually done in the parameter-less [`createStatement()`](https://github.com/neo4j-contrib/neo4j-jdbc/blob/3.0/neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/BoltConnection.java#L138) but not the overloads or the various overloads of `prepareStatement()`. The behavior would then be closer to the HTTP driver.
## Rollback ?

Is it actually an error to rollback an empty transaction anyway?

---

I could actually work on the fix once a consensus has been reached on the solution.

**Comment by glaggia-larus** [comment]
I am investigating on this, I have to admit that I don't have more than few hours today, but I am forced to ask you:

can you to replicate this error with an IT test?

I was looking at the issue on liquigraph but it's not of too much help (for me) in understanding precisely when the error is happening.

What I can briefly say is that the behavior of BOLT and HTTP module **has to be equal in both cases**, for this reason we will investigate this as a bug.
I can also say that I don't think rolling back an empty transaction should be a problem, but it's my personal opinion because I was not able to find any reference to empty transactions in the documentation.
As for the extra nullity check you might be right and I am trying to reproduce an error with a test.

**Comment by glaggia-larus** [comment]
In addition to what I wrote:

**TL;DR**
there is technically no nullity check missing in that method.

---

The connection is created with autocommit set to true as default.
When calling `setAutoCommit(boolean)` on a connection the first case is ALWAYS against an autoCommit true flag. This mean that, trying `setAutoCommit(true)` is a NOOP, and `setAutoCommit(false)` will never enter neither the first `if` nor the `else` clause.
This should be the only case when a transaction is null, from that moment on the transaction will always be not-null as it is created in the second if.

In any case, I will change that part of code to this:

``` java
@Override public void setAutoCommit(boolean autoCommit) throws SQLException {
        if (this.autoCommit != autoCommit) {
            if (this.transaction != null && !this.autoCommit) {
                this.commit();
                this.transaction.close();
            }
            if (this.autoCommit) {
                //Simply restart the transaction
                this.transaction = this.session.beginTransaction();
            }
            this.autoCommit = autoCommit;
        }
    }
```

just to be over protective and formally more precise.
This is also the reason why I was not able to create a bug with my tests.
This is for both Statement and PreparedStatement.

**Comment by fpavageau** [comment]
I've opened PR #63 yesterday, containing integration tests showing the problem in a first commit, and a proposed fix in the second commit, by not throwing exceptions for empty commits or rollbacks.

The main behavior that is expected is still there: throw an exception when trying to commit or roll back a connection with auto-commit on. That's managed by `HttpConnection`, as is all the lifecycle, which `CypherExecutor` should not concern itself with. I had to adjust a few tests which expected `CypherExecutor` to deal with it.

**Comment by fpavageau** [comment]
By the way, here's what [the javadoc for `Connection.commit()` and `Connection.rollback()`](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#commit--) says about the thrown `SQLException`:

> if a database access error occurs, this method is called while participating in a distributed transaction, if this method is called on a closed connection or this Connection object is in auto-commit mode

It says nothing about an empty transaction, so I think the fix respects the contract. It's nicer to be idempotent, too.
